src/ipc-rpc-state-share.ts

```diff
@@ -0,0 +1,256 @@
+  }
+
+  // Returns an opaque string identifying the server's listening
+  // address/path and the secret key used for authentication.
+  public get handle(): string {
+    return JSON.stringify([this.hmacKey.toString('base64'), this.server.address()]);
+  }
+
+  protected async _init(): Promise<void> {
+    this.server = createHTTPServer((req, res) => this._handleRequest(req, res));
+    this.hmacKey = (await promisify(crypto.randomBytes)(16));
+    await this._serverListen();
+  }
+
+  private async _serverListen(): Promise<void> {
+    if (this.cleanupCallback) {
+      throw new Error('RPC server already listening');
+    }
+    let path: string;
+    const name = `${this.serverPathPrefix}_${process.pid}_ipc_${RpcServer.ctr++}_${Math.floor(Math.random() * 10000)}`;
+    if (process.platform === 'win32') {
+      path = `\\\\.\\pipe\\${name}`;
+    } else {
+      const tmpdir = process.env.XDG_RUNTIME_DIR ?? os.tmpdir();
+      path = resolvePath(`${tmpdir}/${name}.sock`);
+    }
+    this.server.listen({ path, exclusive: true });
+    await once(this.server, 'listening');
+    // The race condition in applying chmod after listening is just inherent
+    // to the Node.js API for sockets. We said we'd set the file mode to 0o600
+    // in the design and it probably doesn't hurt to do this either way.
+    await fs.chmod(path, 0o600);
+
+    // In practice, mongosh and Compass may not always know when a connection
+    // state fully closes, so we're adding a process 'exit' listener that cleans
+    // up the file if it hasn't bee closed so far.
+    this.cleanupCallback = () => {
+      if (process.platform !== 'win32') {
+        try {
+          rmSync(path);
+        } catch { /* ignore */ }
+      }
+    };
+    process.on('exit', this.cleanupCallback);
+  }
+
+  private async _handleRequest(req: IncomingMessage, res: ServerResponse) {
+    let response;
+    try {
+      const content = JSON.parse(await readSignedStream(req,
+        req.headers['x-signature'], this.hmacKey));
+
+      response = { status: 200, ...await this.handleRpc(content) };
+    } catch (err) {
+      response = { status: 500, error: err && typeof err === 'object' && 'message' in err ? err.message : String(err) };
+    }
+    res.statusCode = response.status;
+    const payload = new TextEncoder().encode(JSON.stringify(response));
+    res.setHeader('content-type', 'application/json');
+    res.setHeader('content-length', payload.length);
+    res.setHeader('x-signature', sign(payload, this.hmacKey));
+    res.end(payload);
+  }
+
+  public async close(): Promise<void> {
+    this.server.close();
+    await once(this.server, 'close');
+    this.cleanupCallback?.();
+    this.cleanupCallback = undefined;
+  }
+
+  protected abstract handleRpc(content: Record<string, unknown>): Promise<Record<string, unknown>>;
+}
+
+// Client correspondong to RpcServer instances.
+export class RpcClient {
+  private hmacKey: Buffer
```

Comment:
```suggestion
  private hmacKey: Buffer;
```

Reply:
Yeah … prettier would be nice to have here :)